src/coreclr/jit/liveness.cpp

```diff
@@ -2148,7 +2148,15 @@ bool Compiler::fgTryRemoveNonLocal(GenTree* node, LIR::Range* blockRange)
             DISPNODE(node);
 
             node->VisitOperands([](GenTree* operand) -> GenTree::VisitResult {
-                operand->SetUnusedValue();
+                if (!operand->IsValue())
+                {
+                    // Child sets flags, which are consumed by the parent.
+                    assert(operand->gtSetFlags());
```

Comment:
Can you shed some light on what the IR with the new nodes you are adding is going to look like?

It is unfortunate to add new LIR nodes that have operands that are not values. In HIR it might be necessary for ordering purposes, but since ordering in LIR is explicit it would be better to get rid of these operand edges during rationalization, if possible. Rationalization of `GT_CALL` does something similar.

Reply:
This was to prevent the final generated code from putting the result of the compare in a register, which is then unused
eg:
```
\--*  SELECT    ref   
   +--*  EQ        int   
   |  +--*  LCL_VAR   ref    V10 tmp7         
   |  \--*  CNS_INT   ref    null
   +--*  CNS_INT   ref    null
   \--*  LCL_VAR   ref    V34 tmp31        

```
would generate something like:
```
cmp x0, x1
cset x2, eq
csel x3, x4, x5, eq
```
The cset isn't required, as the csel works off the flags. Using GTF_SET_FLAGS seemed the right way of doing that.

Reply:
This seems potentially problematic -- what is going to happen if one of the operands after the `EQ` is also a `SELECT`? Seems like the flags will be trashed.

I can understand why you went this way since `JTRUE` is also doing the same thing. `JTRUE` is a unary node so it is much less prone to the issue, but I would not call what it is doing correct either.

The more usual way to do this is to mark the `EQ` as contained (`setContained()`) and handling it as part of its parent (the `SELECT` in this case). That will essentially reorder the `EQ` with the other operands, for which some interference checking is necessary. You can use the badly named `IsSafeToContainMem` for this (I will probably submit a PR to rename it in the near future).

Reply:
Looking at isContained:
```
// isContained: check whether this tree node is a subcomponent of its parent for codegen purposes
//
// Return Value:
//    Returns true if there is no code generated explicitly for this node.
//    Essentially, it will be rolled into the code generation for the parent.
```

So, then with your suggestion, and using the IR from above:
* The `EQ` node would be skipped at the emit stage (due to being contained)
* The codegen function for the `SELECT` node would need to emit both the `cmp` and `csel` instructions

Is that right?


For the CEQ etc nodes, they'll always end up sandwiched between SELECT node and a standard compare node. And can even end up with multiple of them, eg:
```
*  ASG       int    $VN.Void
+--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
\--*  SELECT    int   
   +--*  CGT       int   
   |  +--*  CNE       int   
   |  |  +--*  LT        int    $102
   |  |  |  +--*  LCL_VAR   int    V02 arg2         u:1 (last use
   |  |  |  \--*  CNS_INT   int    7 $46
   |  |  +--*  LCL_VAR   int    V01 arg1         u:1 $81
   |  |  \--*  CNS_INT   int    10 $45
   |  +--*  LCL_VAR   int    V00 arg0         u:1 $80
   |  \--*  CNS_INT   int    3 $44
   +--*  CNS_INT   int    5 $47
   \--*  LCL_VAR   int    V00 arg0         u:3 $VN.Void
```
generating:
```
IN0001: 000008                    cmp     w2, #7
IN0002: 00000C                    ccmp    w1, #10, z, lt
IN0003: 000010                    ccmp    w0, #3, nzc, ne
IN0004: 000014                    mov     w2, #5
IN0005: 000018                    csel    w0, w2, w0, gt
```

In that instance, I'd be setting the `CGT`, `CNE` and `LT` all as contained.
The code generation `SELECT` node would then need to ensure the `CGT`, `CNE` and `LT` get emitted.

Reply:
> So, then with your suggestion, and using the IR from above:
> 
> * The `EQ` node would be skipped at the emit stage (due to being contained)
> * The codegen function for the `SELECT` node would need to emit both the `cmp` and `csel` instructions
> 
> Is that right?

Exactly.



> For the CEQ etc nodes, they'll always end up sandwiched between SELECT node and a standard compare node. And can even end up with multiple of them, eg:
> 
> ```
> *  ASG       int    $VN.Void
> +--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
> \--*  SELECT    int   
>    +--*  CGT       int   
>    |  +--*  CNE       int   
>    |  |  +--*  LT        int    $102
>    |  |  |  +--*  LCL_VAR   int    V02 arg2         u:1 (last use
>    |  |  |  \--*  CNS_INT   int    7 $46
>    |  |  +--*  LCL_VAR   int    V01 arg1         u:1 $81
>    |  |  \--*  CNS_INT   int    10 $45
>    |  +--*  LCL_VAR   int    V00 arg0         u:1 $80
>    |  \--*  CNS_INT   int    3 $44
>    +--*  CNS_INT   int    5 $47
>    \--*  LCL_VAR   int    V00 arg0         u:3 $VN.Void
> ```
> 
> generating:
> 
> ```
> IN0001: 000008                    cmp     w2, #7
> IN0002: 00000C                    ccmp    w1, #10, z, lt
> IN0003: 000010                    ccmp    w0, #3, nzc, ne
> IN0004: 000014                    mov     w2, #5
> IN0005: 000018                    csel    w0, w2, w0, gt
> ```
> 
> In that instance, I'd be setting the `CGT`, `CNE` and `LT` all as contained. The code generation `SELECT` node would then need to ensure the `CGT`, `CNE` and `LT` get emitted.

Yes, it is ok to recursively contain the nodes and generate them all as part of the parent.

I would like the semantics of all these nodes to be defined in terms of values and not in terms of something they produce or consume from the CPU flags register. The JIT does not model the CPU flags register today in any of its phases and without doing so, producing/consuming values in the CPU flags register is going to introduce (more) unconventional ordering constraints in the IR. I think it rather should be an optimization in the backend that allows the JIT to skip materializing the values when it sees that using the flags register directly is possible. Not doing the optimization should still be correct (for example, every instruction producing 0/1 in some general-purpose register in that case).

I'm not totally sure I understand the semantics of your `CEQ`-family of nodes and I'm not totally sure if they are compatible with this idea. Can you explain it some more? Would it be possible to rephrase them entirely in terms of `SELECT`+ the preexisting compares, and make it a backend optimization that decides when they can be emitted?

Reply:
I've reworked this to use contained nodes, and the logic now feels simpler to follow. The GTF_SET_FLAGS method always felt a little odd. So, totally happy to use contained nodes. The multiple contained nodes works fine too.

>I'm not totally sure I understand the semantics of your CEQ-family of nodes and I'm not totally sure if they are compatible with this idea. Can you explain it some more? Would it be possible to rephrase them entirely in terms of SELECT+ the preexisting compares, and make it a backend optimization that decides when they can be emitted?


Essentially, a CEQ node is a conditional compare equals. It's a standard EQ node that takes in an additional operand. The compare is only taken if this additional operand is true. The other conditions work the same way (CNE, CLT etc).
The SELECT node takes in a condition and two operands. If the condition is true, then select the first operand, else select the second.

Re-using the code from above, we're getting rid the branches from if's with multiple statements.
So:
```
        static void TransformsGtAndAndSingle(uint op1, uint op2, uint op3) {
            if (op1 > 3 && op2 != 10 && op3 < 7) {
                op1 = 5;
            }
            Consume(op1, op2);
        }
```
Becomes the standard IR:
```
------------ BB01 [000..004) -> BB05 (cond), preds={} succs={BB02,BB05}

***** BB01
STMT00000 ( 0x000[E-] ... 0x002 )
N004 (  5,  6) [000003] -----------                         *  JTRUE     void   $VN.Void
N003 (  3,  4) [000002] N------N-U-                         \--*  LE        int    $100
N001 (  1,  1) [000000] -----------                            +--*  LCL_VAR   int    V00 arg0         u:1 $80
N002 (  1,  2) [000001] -----------                            \--*  CNS_INT   int    3 $44

------------ BB02 [004..009) -> BB05 (cond), preds={BB01} succs={BB03,BB05}

***** BB02
STMT00003 ( 0x004[E-] ... 0x007 )
N004 (  5,  6) [000011] -----------                         *  JTRUE     void   $VN.Void
N003 (  3,  4) [000010] J------N---                         \--*  EQ        int    $101
N001 (  1,  1) [000008] -----------                            +--*  LCL_VAR   int    V01 arg1         u:1 $81
N002 (  1,  2) [000009] -----------                            \--*  CNS_INT   int    10 $45

------------ BB03 [009..00D) -> BB05 (cond), preds={BB02} succs={BB04,BB05}

***** BB03
STMT00004 ( 0x009[E-] ... 0x00B )
N004 (  7,  7) [000015] -----------                         *  JTRUE     void   $VN.Void
N003 (  5,  5) [000014] N------N-U-                         \--*  GE        int    $102
N001 (  3,  2) [000012] -----------                            +--*  LCL_VAR   int    V02 arg2         u:1 (last use) $82
N002 (  1,  2) [000013] -----------                            \--*  CNS_INT   int    7 $46

------------ BB04 [00D..010), preds={BB03} succs={BB05}

***** BB04
STMT00005 ( 0x00D[E-] ... 0x00E )
N003 (  1,  3) [000018] -A------R--                         *  ASG       int    $VN.Void
N002 (  1,  1) [000017] D------N---                         +--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
N001 (  1,  2) [000016] -----------                         \--*  CNS_INT   int    5 $47

------------ BB05 [010..018) (return), preds={BB01,BB02,BB03,BB04} succs={}

***** BB05
STMT00006 ( ??? ... ??? )
N005 (  0,  0) [000021] -A------R--                         *  ASG       int    $VN.Void
N004 (  0,  0) [000019] D------N---                         +--*  LCL_VAR   int    V00 arg0         d:2 $VN.Void
N003 (  0,  0) [000020] -----------                         \--*  PHI       int    $140
N001 (  0,  0) [000023] ----------- pred BB04                  +--*  PHI_ARG   int    V00 arg0         u:3 $47
N002 (  0,  0) [000022] ----------- pred BB01                  \--*  PHI_ARG   int    V00 arg0         u:1 $80

***** BB05
STMT00001 ( 0x010[E-] ... 0x017 )
N003 ( 16,  6) [000006] --CXG------                         *  CALL      void   CSharpTutorials.Program.Consume $VN.Void
N001 (  1,  1) [000004] ----------- arg0 in x0              +--*  LCL_VAR   int    V00 arg0         u:2 (last use) $140
N002 (  1,  1) [000005] ----------- arg1 in x1              \--*  LCL_VAR   int    V01 arg1         u:1 (last use) $81

***** BB05
STMT00002 ( 0x017[E-] ... ??? )
N001 (  0,  0) [000007] -----------                         *  RETURN    void   $VN.Void
```

Blocks 3 and 4 get if converted to:
```
***** BB03
STMT00004 ( 0x009[E-] ... 0x00B )
N001 (  0,  0) [000029] -----------                         *  NOP       void  

------------ BB04 [00D..010), preds={BB03} succs={BB05}

***** BB04
STMT00005 ( 0x00D[E-] ... 0x00E )
N008 (  1,  3) [000018] -A------R--                         *  ASG       int    $VN.Void
N007 (  1,  1) [000017] D------N---                         +--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
N006 ( 10,  8) [000028] -----------                         \--*  SELECT    int   
N003 (  8,  5) [000025] -------N-U-                            +--*  LT        int    $102
N001 (  3,  2) [000026] -----------                            |  +--*  LCL_VAR   int    V02 arg2         u:1 (last use) $82
N002 (  1,  2) [000027] -----------                            |  \--*  CNS_INT   int    7 $46
N004 (  1,  2) [000016] -----------                            +--*  CNS_INT   int    5 $47
N005 (  1,  1) [000024] -----------                            \--*  LCL_VAR   int    V00 arg0         u:3 $VN.Void
```

Then blocks 2, 3 and 4 get converted:
```
***** BB02
STMT00003 ( 0x004[E-] ... 0x007 )
N001 (  0,  0) [000031] -----------                         *  NOP       void  

------------ BB03 [009..00D), preds={BB02} succs={BB04}

***** BB03
STMT00004 ( 0x009[E-] ... 0x00B )
N001 (  0,  0) [000029] -----------                         *  NOP       void  

------------ BB04 [00D..010), preds={BB03} succs={BB05}

***** BB04
STMT00005 ( 0x00D[E-] ... 0x00E )
N011 (  1,  3) [000018] -A------R--                         *  ASG       int    $VN.Void
N010 (  1,  1) [000017] D------N---                         +--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
N009 ( 10,  8) [000028] N----------                         \--*  SELECT    int   
N006 ( 10,  8) [000030] -----------                            +--*  CNE       int   
N003 (  8,  5) [000025] -------N-U-                            |  +--*  LT        int    $102
N001 (  3,  2) [000026] -----------                            |  |  +--*  LCL_VAR   int    V02 arg2         u:1 (last use) $82
N002 (  1,  2) [000027] -----------                            |  |  \--*  CNS_INT   int    7 $46
N004 (  1,  1) [000008] -----------                            |  +--*  LCL_VAR   int    V01 arg1         u:1 $81
N005 (  1,  2) [000009] -----------                            |  \--*  CNS_INT   int    10 $45
N007 (  1,  2) [000016] -----------                            +--*  CNS_INT   int    5 $47
N008 (  1,  1) [000024] -----------                            \--*  LCL_VAR   int    V00 arg0         u:3 $VN.Void
```

Then finally blocks 1,2,3,4:
```
------------ BB01 [000..004), preds={} succs={BB02}

***** BB01
STMT00000 ( 0x000[E-] ... 0x002 )
N001 (  0,  0) [000033] -----------                         *  NOP       void  

------------ BB02 [004..009), preds={BB01} succs={BB03}

***** BB02
STMT00003 ( 0x004[E-] ... 0x007 )
N001 (  0,  0) [000031] -----------                         *  NOP       void  

------------ BB03 [009..00D), preds={BB02} succs={BB04}

***** BB03
STMT00004 ( 0x009[E-] ... 0x00B )
N001 (  0,  0) [000029] -----------                         *  NOP       void  

------------ BB04 [00D..010), preds={BB03} succs={BB05}

***** BB04
STMT00005 ( 0x00D[E-] ... 0x00E )
N014 (  1,  3) [000018] -A------R--                         *  ASG       int    $VN.Void
N013 (  1,  1) [000017] D------N---                         +--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
N012 ( 10,  8) [000028] -----------                         \--*  SELECT    int   
N009 ( 12, 11) [000032] -----------                            +--*  CGT       int   
N006 ( 10,  8) [000030] -----------                            |  +--*  CNE       int   
N003 (  8,  5) [000025] -------N-U-                            |  |  +--*  LT        int    $102
N001 (  3,  2) [000026] -----------                            |  |  |  +--*  LCL_VAR   int    V02 arg2         u:1 (last use) $82
N002 (  1,  2) [000027] -----------                            |  |  |  \--*  CNS_INT   int    7 $46
N004 (  1,  1) [000008] -----------                            |  |  +--*  LCL_VAR   int    V01 arg1         u:1 $81
N005 (  1,  2) [000009] -----------                            |  |  \--*  CNS_INT   int    10 $45
N007 (  1,  1) [000000] -----------                            |  +--*  LCL_VAR   int    V00 arg0         u:1 $80
N008 (  1,  2) [000001] -----------                            |  \--*  CNS_INT   int    3 $44
N010 (  1,  2) [000016] -----------                            +--*  CNS_INT   int    5 $47
N011 (  1,  1) [000024] -----------                            \--*  LCL_VAR   int    V00 arg0         u:3 $VN.Void

```

Eventually generating:
```
G_M31930_IG01:        ; func=00, offs=000000H, size=0008H, bbWeight=1    PerfScore 1.50, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref, nogc <-- Prolog IG

IN000b: 000000                    stp     fp, lr, [sp,#-16]!
IN000c: 000004                    mov     fp, sp

G_M31930_IG02:        ; offs=000008H, size=0028H, bbWeight=1    PerfScore 8.00, gcrefRegs=0000 {}, byrefRegs=0000 {}, BB01 [0000], byref

IN0001: 000008                    mov     w3, #5
IN0002: 00000C                    cmp     w2, #7
IN0003: 000010                    ccmp    w1, #10, z, lt
IN0004: 000014                    ccmp    w0, #3, nzc, ne
IN0005: 000018                    csel    w0, w3, w0, gt
IN0006: 00001C                    movz    x2, #0x7a38      // code for CSharpTutorials.Program:Consume
IN0007: 000020                    movk    x2, #0xb771 LSL #16
IN0008: 000024                    movk    x2, #0xffff LSL #32
IN0009: 000028                    ldr     x2, [x2]
IN000a: 00002C                    blr     x2

G_M31930_IG03:        ; offs=000030H, size=0008H, bbWeight=1    PerfScore 2.00, epilog, nogc, extend

IN000d: 000030                    ldp     fp, lr, [sp],#16
IN000e: 000034                    ret     lr
```

Essentially: Compare `w2, #7`. if they are `lt` then compare `w1, #10`. If they are `ne` then compare `w0, #3`. If they were `gt` then `w0=w3`, else `w0=w0`.

(The 3rd arg of each ccmp is ensuring the failure cases work).

Hope that's a bit clearer.

Reply:
> I've reworked this to use contained nodes, and the logic now feels simpler to follow. The GTF_SET_FLAGS method always felt a little odd. So, totally happy to use contained nodes. The multiple contained nodes works fine too.

Great, happy to hear that it worked out!

> Essentially: Compare w2, #7. if they are lt then compare w1, #10. If they are ne then compare w0, #3. If they were gt then w0=w3, else w0=w0.

So if I understand correctly `CEQ(A, B, C)` is equivalent to `A & (B == C)` and `CLT(A, B, C)` is equivalent `A & (B < C)`, and so forth?

Reply:
So, in that scenario, we'd still have a three op SELECT node.

This graph:
```
*  ASG       int    $VN.Void
+--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
\--*  SELECT    int   
   +--*  CGT       int   
   |  +--*  CNE       int   
   |  |  +--*  LT        int    $102
   |  |  |  +--*  LCL_VAR   int    V02 arg2         u:1 (last use) $82
   |  |  |  \--*  CNS_INT   int    7 $46
   |  |  +--*  LCL_VAR   int    V01 arg1         u:1 $81
   |  |  \--*  CNS_INT   int    10 $45
   |  +--*  LCL_VAR   int    V00 arg0         u:1 $80
   |  \--*  CNS_INT   int    3 $44
   +--*  CNS_INT   int    5 $47
   \--*  LCL_VAR   int    V00 arg0         u:3 $VN.Void
```
Is instead:
```
*  ASG       int    $VN.Void
+--*  LCL_VAR   int    V00 arg0         d:3 $VN.Void
\--*  SELECT    int   
   +--*  AND
   |  +--*  AND
   |  |  +--*  NE
   |  |  |  +--*  LCL_VAR   int    V01 arg1         u:1 $81
   |  |  |  \--*  CNS_INT   int    10 $45
   |  |  \--*  LT        int    $102
   |  |     +--*  LCL_VAR   int    V02 arg2         u:1 (last use) $82
   |  |     \--*  CNS_INT   int    7 $46
   |  \--*  GT       int   
   |     +--*  LCL_VAR   int    V00 arg0         u:1 $80
   |     \--*  CNS_INT   int    3 $44
   +--*  CNS_INT   int    5 $47
   \--*  LCL_VAR   int    V00 arg0         u:3 $VN.Void

```

* That does have an advantage with constant folding etc - all the existing code will just work. It should be easy put the If conversion pass before constant propagation / CSE etc without any big code changes. The footprint of the patch across the whole jit is smaller. 
* The graph is now two nodes larger (as each conditional compare is now 2 nodes).
* Tracking whether the condition node of the SELECT is valid is harder. Previously it just needed to ensure the condition node was a compare or conditional compare node (eg EQ or CEQ). Now, the SELECT node is having to check if the entire sub tree is a collection of ANDed together compares - as it needs to ensure it ends up with a condition flag set at the end.

Reply:
> So, in that scenario, we'd still have a three op SELECT node.

Right, SELECT should stay. I like this more since it means we do not end up with two different ways of representing things. That would be more reasonable if there were downsides to one of the methods of evaluating this, but it seems like using the conditional compares is always beneficial here.

> Tracking whether the condition node of the SELECT is valid is harder. Previously it just needed to ensure the condition node was a compare or conditional compare node (eg EQ or CEQ). Now, the SELECT node is having to check if the entire sub tree is a collection of ANDed together compares - as it needs to ensure it ends up with a condition flag set at the end.

I agree that tracking whether the whole thing can be contained/optimized becomes a bit harder. I just want to point out that the SELECT should be valid regardless of what the condition operand contains (of course the type must be correct). We should not be saying that things like `SELECT(CNS_INT 0, A, B)` or `SELECT(LCL_VAR C, A, B)` are invalid IR, or relying on the fact that some nodes produce condition flags in the flags register (the latter would require us to model the CPU flags at a level that we are not doing today).

In practice this means that in some cases lowering may notice that the condition _cannot_ be contained in the `SELECT`. In that case the codegen for `SELECT` will need to generate an additional compare instruction using the register that the condition was evaluated into.

Reply:
Code now uses contained nodes instead of relying on flags.

I'll look at replacing CEQ etc with AND.

Reply:
Latest version of the patch uses AND/CMP chains instead of the CEQ, CNE etc nodes. Main changes are in lower.cpp and codegenarm64.cpp. The most awkward part here was deciding what combinations of contained and not contained are allowed in a chain - eventually I went with the left side of the AND can only be contained if the right side is fully contained. This ensures we generate an unbroken chain of ccmp nodes going backwards from the select node.

CEQ, CNE etc nodes are no longer used. I've not yet removed the support for them - this was to keep the diffs smaller and make it easier to review. I'll remove support for them if we're happy the AND/CMP chains are the better approach.

Reply:
Something else that becomes awkward when we have AND Compare chains:

Op2 of an Arm64 `CMP` instruction can be contained if it's a funky 12bit immediate.
Op2 of an Arm64 `CCMP` instruction can be contained if it's a 5bit immediate.
(see `emitIns_valid_imm_for_cmp` and `emitIns_valid_imm_for_ccmp`).

When checking if the Op2 of any Compare can be contained, I need to figure out if it's part of a chain and whether it'll generate a `CMP` or `CCMP`.

Eg: In `SELECT(AND(AND(EQ, LT),NE), X, Y)`
The NE will generate a CMP and the EQ and LT will generate a CMP.

Reply:
It might be simpler to just clear containment on the immediate if you convert it to a ccmp and it is no longer containable. You can use `ClearContained` for that.

Reply:
AND / CMP chain changes added.

To get the immediate correct, I assume a contained CMP will generate a CCMP.  When calculating the chains, I then have a little bit of code which corrects the first contained CMP (due to the first CMP in a chain always being generated as a CMP)